### ISSUE TITLE
Mouse up outside map will cause "canvas-drag" event to stick

### ISSUE THREAD

**Issue by LOMS** [openseadragon/openseadragon#697]
As subject says, if you dragging map and your cursor is outside map, when you release mouse button - map will stick to your cursor until you reload current page.
This works with firefox. Try CTRL + D and drag your mouse into there.

**Comment by iangilman** [comment]
@msalsbery is there anything we can do about this?

**Comment by iangilman** [comment]
@LOMS Do you mean dragging outside the window? Or just outside the viewer? Either way, I can't reproduce it on http://openseadragon.github.io/ with Firefox Mac. What site/browser/OS combo are you using?

**Comment by mzeinstra** [comment]
I also experience this problem with a new platform that allows embedding IIIF files with OpenSeaDragon. It uses iFrames to embed the OpenSeaDragon. When you put the player in an iFrame, start dragging and your mouse ends up outside the frame the problem occurs.

I've made an example at http://test.kl.nl/osd/test.html using http://iangilman.com/openseadragon/rdio/ to fill my test iFrame. 

Steps to reproduce:
* Load http://test.kl.nl/osd/test.html
* Drag the image, keep draging until the mouse pointer is outside the frame
* Return the mouse to the frame
* canvas-drag sticks to mouse, only reload solves the issue

Work arounds?

**Comment by LOMS** [comment]
Sorry for delayed answer, completely forget about the ticket.
Way to reproduce for me: Latest Firefox + plugin "All in One sidebar". If I'm accidently drop mouse over AIOSidebar, while dragging image, canvas-drag will stick to mouse.

**Comment by iangilman** [comment]
Related tickets: #427, #464, #579. 

@msalsbery Any thoughts? 

@mzeinstra @LOMS Any interest in looking into a fix?

**Comment by msalsbery** [comment]
I still don't have a solution that works for iframes across browsers but it's a priority on my TODO list. 

Really, I had hoped this one was forgotten ;)

**Comment by LOMS** [comment]
@iangilman what do you mean "looking into fix"? If you ask me - just raise "canvas-drag-end" on each mouseup. But since I'm not too much skilled in JS, this could be undistilled, stupid advice.

**Comment by msalsbery** [comment]
The advice is sound. The reality of browser implementations of the desired events is where things get, um, inconsistent.

**Comment by mzeinstra** [comment]
I had succes (Chrome, Safari and Firefox on Mac OS X 10.11) with:
```
		document.documentElement.onmouseout = function ()
		{
			window.viewer.canvas-drag-end();
		}
```

It does give an error in the console when you are not dragging, but that should be fixable with an if-statement somewhere. Not a prof. developer btw. :) 

See http://test.kl.nl/osd/solution/Rijksmuseum_SK-C-5.html
(some images buttons broke as I hacked this together real quick)

**Comment by LOMS** [comment]
@mzeinstra nope, doesn't work. First time it has blocked canvas-drag from firing, second time it didn't work at all.
And there are error in console: ReferenceError: drag is not defined

**Comment by josmilmed** [comment]
I temporary solved this with the following modifications (MouseTracker.js, lines 209 and 211):
```javascript
mouseleave: function ( event ) { onMouseLeave( _this, event ); onMouseUp( _this, event ); }, 
mouseout:    function ( event ) { onMouseOut( _this, event ); onMouseUp( _this, event ); },
```
I know it is a dirty solution, but still works for me.

**Comment by willbuckingham** [comment]
+1 This issue.
I attempted @mzeinstra 's solution, but it does not work in IE, I believe because it is trying to call an event as a function and IE does not appreciate the hyphens.

I tried @josmilmed 's solution and it seems to work in non-IE browsers.  I then attempted to modify the "pointer.." events with similar logic.  And adding various pointer up functions to the over or out function does not correct the issue.

I get a partial fix with adding "onPointerUpCaptured" to the "pointerdown", so that when I reclick in the viewer after moving back over the iframe, I can resume normal behavior without it being stuck to the mouse, and not need to refresh.  However, it would be nice if there was a proper fix crossbrowser to handle mousing out.

Let me know if my issue and attempted solutions are unclear, im writing in a rush.
Thanks,

**Comment by iangilman** [comment]
@msalsbery Any more thoughts on this? 

@willbuckingham Any interest in working on a patch for this? Looks like it's up in the air. 

Note that https://github.com/openseadragon/openseadragon/pull/790 addresses the iframe scenario related to this.

**Comment by jrhoads** [comment]
Any updates on this?  I am experiencing the same issue.

This is a great project.
Let me know if there's anything I can do to be helpful.

**Comment by willbuckingham** [comment]
I have not tested with the most recent release (I am still using 2.2.0), but I added this code to my page in order to not have issues:

`document.addEventListener("mouseleave", function(evt){
    viewer.setMouseNavEnabled(false);
    viewer.setMouseNavEnabled(true)
});`

It's probably technically a hack, but it helped my issues and might help yours.

**Comment by iangilman** [comment]
@jrhoads No news here... it seems that no one who was looking at a solution has made a patch. Would you be interested in making a patch? It seems there are a number of strategies to try listed above.

**Comment by leesei** [comment]
I'm also hit by this issue in Chrome, only @willbuckingham's solution works for me.
However this is adding events to `document`, should we add it in document instead?
Or we can detect being in `iframe` before adding the event.

**Comment by willbuckingham** [comment]
@leesei There are a couple checks you could perform to decide if you are in an iframe before you add the handlers, ex:

`var isInIframe = (window.location != window.parent.location) ? true : false;`

I added to document since my player was edge to edge with my webpage, if your circumstances are different perhaps you can do something else.

And while I called it a hack, as I do believe it would be best in OpenSeadragon handled this itself (admittedly, I am not sure how to best add this support to OSD), it is relying on legitimate method calls, they just happen to protect against it getting stuck.

And good call on listening to both leave AND out events.

**Comment by leesei** [comment]
As it turns out, toggling MouseNav will mess up my other features.
I think I have to limit the iframe to an a edge to edge embedding.

**Comment by josmilmed** [comment]
I finally solved the issue in the following way:

**1. Add this method to MouseTracker class:**

$.MouseTracker.resetAllMouseTrackers = function(){
    MOUSETRACKERS.forEach(function(mt){
        mt.setTracking(false);
        mt.setTracking(true);
    });
}

**2. Make the following modification to onCanvasExit function, in Viewer:**

function onCanvasExit( event ) {
    $.MouseTracker.resetAllMouseTrackers(); // <== ADD THIS LINE

    this.raiseEvent( 'canvas-exit', {
        tracker: event.eventSource,
        pointerType: event.pointerType,
        position: event.position,
        buttons: event.buttons,
        pointers: event.pointers,
        insideElementPressed: event.insideElementPressed,
        buttonDownAny: event.buttonDownAny,
        originalEvent: event.originalEvent
    });
}

A bit dirty, but works.

**Comment by msalsbery** [comment]
If it works everywhere then there's nothing dirty about it at all :)

Thanks for the effort on this everyone!

**Comment by iangilman** [comment]
@josmilmed Awesome... if @msalsbery likes it, I like it. Great to have this finally fixed! Can you please make a pull request?

**Comment by josmilmed** [comment]
Sorry, I forgot that you may have inactive mousetrackers, so the best option is replacing:

$.MouseTracker.resetAllMouseTrackers = function(){
MOUSETRACKERS.forEach(function(mt){
mt.setTracking(false);
mt.setTracking(true);
});
}

with:

$.MouseTracker.resetAllMouseTrackers = function(){
MOUSETRACKERS.forEach(function(mt){
if (mt.isTracking()){
mt.setTracking(false);
mt.setTracking(true);
}
});
}

**Comment by iangilman** [comment]
Good call :)

**Comment by josmilmed** [comment]
Pull request done!

**Comment by msalsbery** [comment]
What platforms has this been tested on?

**Comment by josmilmed** [comment]
It was tested on Windows

**Comment by msalsbery** [comment]
Edge? Excellent, good enough for me ;) I'll test on Mac/Windows across Edge, IE, Chrome, Safari, FF whether Ian merges the PR or not. Thank you!

**Comment by iangilman** [comment]
@josmilmed Awesome, thank you!

@msalsbery Please let me know how the testing goes! Thank you for doing it :)

**Comment by iangilman** [comment]
Fixed by #1133
